### PR TITLE
extend dhcp timeout on failure

### PR DESCRIPTION
**Pull request by Eslam-Nawara**
### Description

extend dhcp timeout on failure

### Related Issues

- #2429 

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

### CODE REVIEW

File: pkg/network/dhcp/dhcp.go [hunk 1]
```diff
@@ -43,22 +44,31 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
...
 	var stderr bytes.Buffer
 
-	cmd.Stdout = &stdout
-	cmd.Stderr = &stderr
-	if err := cmd.Run(); err != nil {
+	t := 1
+	check := func() error {
+		args := []string{
+			"-i", inf, // the interface to prope
+			"-q",      // exist once lease is optained
```
Comment (xmonader) [inline pkg/network/dhcp/dhcp.go]: Exit after obtaining lease

File: pkg/network/dhcp/dhcp.go [hunk 2]
```diff
@@ -43,22 +44,31 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
...
+			"-f",      // foreground
+			"-T", "1", // every second
```
Comment (xmonader) [inline pkg/network/dhcp/dhcp.go]: This pauses between the packets timeout

File: pkg/network/dhcp/dhcp.go [hunk 3]
```diff
@@ -43,22 +44,31 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
...
+			"-t", fmt.Sprint(t), // send 't' dhcp queries
+			"-s", "/usr/share/udhcp/probe.script", // use the prope script
```
Comment (xmonader) [inline pkg/network/dhcp/dhcp.go]: probe

File: pkg/network/dhcp/dhcp.go [hunk 4]
```diff
@@ -43,22 +44,31 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
...
-	cmd.Stderr = &stderr
-	if err := cmd.Run(); err != nil {
+	t := 1
+	check := func() error {
+		args := []string{
+			"-i", inf, // the interface to prope
+			"-q",      // exist once lease is optained
+			"-f",      // foreground
+			"-T", "1", // every second
+			"-t", fmt.Sprint(t), // send 't' dhcp queries
```
Comment (xmonader) [inline pkg/network/dhcp/dhcp.go]: dhcp queries -> dhcp discover packets

File: pkg/network/dhcp/dhcp.go [hunk 5]
```diff
@@ -43,22 +44,31 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
...
+
+		cmd := exec.CommandContext(ctx, "udhcpc", args...)
+		cmd.Stdout = &stdout
+		cmd.Stderr = &stderr
+
+		t += 5
+		return cmd.Run()
+	}
+
+	if err := backoff.Retry(check, backoff.NewExponentialBackOff()); err != nil {
```
Comment (xmonader) [inline pkg/network/dhcp/dhcp.go]: should we set max retries here?

File: pkg/network/dhcp/dhcp.go [hunk 6]
```diff
@@ -43,22 +44,31 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
...
 		return output, errors.Wrapf(err, "failed to prope interface '%s': %s", inf, stderr.String())
```
Comment (xmonader) [inline pkg/network/dhcp/dhcp.go]: failed to probe interface

File: pkg/network/dhcp/dhcp.go [hunk 7]
```diff
@@ -43,22 +44,31 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
 func Probe(ctx context.Context, inf string) (output ProbeOutput, err error) {
 	// use udhcpc to prope the interface.
```
Comment (xmonader) [inline pkg/network/dhcp/dhcp.go]: probe

File: pkg/network/dhcp/dhcp.go [hunk 8]
```diff
@@ -43,22 +44,31 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
...
 	// this depends on that the interface is UP
```
Comment (xmonader) [inline pkg/network/dhcp/dhcp.go]: should we have a guard here so we don't bother if the interface is down?

File: pkg/network/dhcp/dhcp.go (hunk 7 above)
Comment (iwanbk) [reply pkg/network/dhcp/dhcp.go]: consider to add `misspell` to the golangci-lint config https://golangci-lint.run/usage/linters/#misspell

File: pkg/network/dhcp/dhcp.go [hunk 9]
```diff
@@ -41,25 +42,47 @@ func (p *ProbeOutput) IPNet() (*net.IPNet, error) {
...
+
+		t += 5
+		return cmd.Run()
+	}
+
+	if err := backoff.Retry(
+		check,
+		backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetries),
+	); err != nil {
+		return output, errors.Wrapf(err, "failed to probe interface '%s': %s", inf, stderr.String())
```
Comment (muhamadazmy) [inline pkg/network/dhcp/dhcp.go:85]: The udhcpc command already retries and you can control how many times and delay between sends via it's argument (the `-t` and `-T`) options I am wondering if it's necessary to implement another retry mechanism on top

**Review by muhamadazmy — CHANGES_REQUESTED** [review]
I think this is the wrong place to implement retries. If you check the `internet` binary you will see that it implements its own retry on top of this `prob` The probe is has it's own retry mechansim with udhcpc command.

It's up to the user of the probe to decide. for example `internet` never gives up until it has an ip. Other users might give up or retry periodically that's where they can retry the Probe!

**Pull request closed without merging**